### PR TITLE
Add integration tests for controllers

### PR DESCRIPTION
**Pull request by Anna-Kudiakova**

### CODE REVIEW

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<dataset>
+    <medicines id="2021111201" title="MedicineControllerIT_name1"></medicines>
+    <medicines id="2021111202" title="MedicineControllerIT_name2"></medicines>
+</dataset>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml:4]: leave blank line here, pls

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<dataset>
+    <pharmacies id="2021111201" name="PharmacyControllerIT_name1" medicine_link_template ="PharmacyControllerIT_link1"></pharmacies>
+    <pharmacies id="2021111202" name="PharmacyControllerIT_name2" medicine_link_template ="PharmacyControllerIT_link2"></pharmacies>
+    <medicines id="2021111201" title="MedicineControllerIT_name1"></medicines>
+    <medicines id="2021111202" title="MedicineControllerIT_name2"></medicines>
+    <prices pharmacy_id="2021111201" medicine_id="2021111201" price="150" external_id ="2021111201"></prices>
+    <prices pharmacy_id="2021111202" medicine_id="2021111202" price="200" external_id ="2021111202"></prices>
+</dataset>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:8]: blank line ^-^

**Review by MuchosChuchos — COMMENTED** [review]
the first thing that caught my eye

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml (hunk 1 above)
Comment (Anna-Kudiakova) [reply api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml]: Done!

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml (hunk 2 above)
Comment (Anna-Kudiakova) [reply api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml]: done=)

**Review by SofiiaOlenyn — APPROVED** [review]

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java [hunk 3]
```diff
@@ -29,13 +33,23 @@
     private MockMvc mockMvc;
     private DatabaseDataSourceConnection dataSourceConnection;
 
+    private final String uri = "/pharmacies";
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java:36]: pls make it in upper-case

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java [hunk 4]
```diff
@@ -54,6 +68,79 @@ public void findAllPharmacies_findIds_ok() throws Exception {
         }
     }
 
+    @Test
+    public void findPharmacyById_ok() throws Exception {
+        final int pharmacyId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java:73]: make it non-final or upper-case, please

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java [hunk 5]
```diff
@@ -54,6 +68,79 @@ public void findAllPharmacies_findIds_ok() throws Exception {
...
+                    .andExpect(jsonPath("$.name").value("PharmacyControllerIT_name1"))
+                    .andExpect(jsonPath("$.medicineLinkTemplate").value("PharmacyControllerIT_link1"));
+        } finally {
+            this.dataSourceConnection.close();
+        }
+    }
+
+    @Test
+    public void findPharmacyById_isNotFound() throws Exception {
+        final int pharmacyId = 2021111203;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java:88]: make it non-final or upper-case, pls

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java [hunk 6]
```diff
@@ -54,6 +68,79 @@ public void findAllPharmacies_findIds_ok() throws Exception {
...
+                    .andExpect(jsonPath("$.medicineLinkTemplate").value("PharmacyControllerIT_link3"));
+        } finally {
+            this.dataSourceConnection.close();
+        }
+
+    }
+
+    @Test
+    void updatePharmacyById_isOk() throws Exception {
+        final int pharmacyId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java:118]: pls, make it non-final or upper-case

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java [hunk 7]
```diff
@@ -54,6 +68,79 @@ public void findAllPharmacies_findIds_ok() throws Exception {
...
+                    .andExpect(MockMvcResultMatchers.status().isOk())
+                    .andExpect(jsonPath("$.name").value("new-name"));
+        } finally {
+            this.dataSourceConnection.close();
+        }
+    }
+
+    @Test
+    void deletePharmacyById_isNoContent() throws Exception {
+        final int pharmacyId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT.java:134]: non-final or upper

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@ActiveProfiles("test")
+public class PriceControllerIT {
+
+    private MockMvc mockMvc;
+    private DatabaseDataSourceConnection dataSourceConnection;
+
+    private final String uri = "/prices";
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:36]: non-final or upper-case

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                    .andExpect(jsonPath("$[*].externalId",
+                            Matchers.hasItems("2021111201","2021111202")));
+        } finally {
+            this.dataSourceConnection.close();
+        }
+    }
+
+    @Test
+    public void findPriceById_ok() throws Exception {
+        final int pharmacyId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:73]: upper-case or non-final

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        final int medicineId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:74]: non-final or upper-case pls

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+                    .andExpect(jsonPath("$.price").value("150.0"))
+                    .andExpect(jsonPath("$.externalId").value("2021111201"));
+        } finally {
+            this.dataSourceConnection.close();
+        }
+    }
+
+    @Test
+    public void findPriceById_isNotFound() throws Exception {
+        final int pharmacyId = 2021110703;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:90]: take look

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        final int medicineId = 2021110703;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:91]: take look pls

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                            + "/pharmacyId/{pharmacyId}/medicineId/{medicineId}", pharmacyId, medicineId))
+                    .andExpect(MockMvcResultMatchers.status().isNotFound());
+        } finally {
+            this.dataSourceConnection.close();
+        }
+    }
+
+    @Test
+    void updatePriceById_isOk() throws Exception {
+        final int pharmacyId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:105]: take look pls

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+        final int medicineId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:106]: same as prev

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+                    .andExpect(MockMvcResultMatchers.status().isOk())
+                    .andExpect(jsonPath("$.price").value("180.0"));
+        } finally {
+            this.dataSourceConnection.close();
+        }
+    }
+
+    @Test
+    void deletePriceById_isNoContent() throws Exception {
+        final int pharmacyId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:123]: another one =)

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+        final int medicineId = 2021111201;
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:124]: take look pls

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml [hunk 17]
```diff
@@ -0,0 +1,4 @@
+<dataset>
+    <medicines id="2021111201" title="MedicineControllerIT_name1"></medicines>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml:2]: use non-container tags here

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml [hunk 18]
```diff
@@ -0,0 +1,4 @@
...
+    <medicines id="2021111202" title="MedicineControllerIT_name2"></medicines>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/MedicineControllerIT_dataset.xml:3]: the same as prev

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT_dataset.xml [hunk 19]
```diff
@@ -1,4 +1,4 @@
 <dataset>
-    <pharmacies id="2021102101" name="PharmacyControllerIT_name1"></pharmacies>
-    <pharmacies id="2021102102" name="PharmacyControllerIT_name2"></pharmacies>
+    <pharmacies id="2021111201" name="PharmacyControllerIT_name1" medicine_link_template ="PharmacyControllerIT_link1"></pharmacies>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT_dataset.xml:2]: non-container tag

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT_dataset.xml [hunk 20]
```diff
@@ -1,4 +1,4 @@
...
+    <pharmacies id="2021111202" name="PharmacyControllerIT_name2" medicine_link_template ="PharmacyControllerIT_link2"></pharmacies>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PharmacyControllerIT_dataset.xml:3]: non-container tag pls ))

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 21]
```diff
@@ -0,0 +1,8 @@
+<dataset>
+    <pharmacies id="2021111201" name="PharmacyControllerIT_name1" medicine_link_template ="PharmacyControllerIT_link1"></pharmacies>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:2]: convert to non-container tag pls

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 22]
```diff
@@ -0,0 +1,8 @@
...
+    <pharmacies id="2021111202" name="PharmacyControllerIT_name2" medicine_link_template ="PharmacyControllerIT_link2"></pharmacies>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:3]: the same as prev.

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 23]
```diff
@@ -0,0 +1,8 @@
...
+    <medicines id="2021111201" title="MedicineControllerIT_name1"></medicines>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:4]: same

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 24]
```diff
@@ -0,0 +1,8 @@
...
+    <medicines id="2021111202" title="MedicineControllerIT_name2"></medicines>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:5]: same

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 25]
```diff
@@ -0,0 +1,8 @@
...
+    <prices pharmacy_id="2021111201" medicine_id="2021111201" price="150" external_id ="2021111201"></prices>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:6]: same

File: api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml [hunk 26]
```diff
@@ -0,0 +1,8 @@
...
+    <prices pharmacy_id="2021111202" medicine_id="2021111202" price="200" external_id ="2021111202"></prices>
```
Comment (MuchosChuchos) [inline api/src/test/resources/com/eleks/academy/pharmagator/controllers/PriceControllerIT_dataset.xml:7]: same

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java (hunk 9 above)
Comment (MuchosChuchos) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:73]: DRY. You can make it global

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java (hunk 10 above)
Comment (MuchosChuchos) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:74]: DRY.))))

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+        try {
+            DatabaseOperation.REFRESH.execute(this.dataSourceConnection, readDataset());
+
+            this.mockMvc.perform(MockMvcRequestBuilders.get(uri
+                            + "/pharmacyId/{pharmacyId}/medicineId/{medicineId}", pharmacyId, medicineId))
+                    .andExpect(MockMvcResultMatchers.status().isOk())
+                    .andExpect(jsonPath("$.price").value("150.0"))
+                    .andExpect(jsonPath("$.externalId").value("2021111201"));
```
Comment (MuchosChuchos) [inline api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java:82]: this value is stored in a variable, hence you can not use string literal.

**Review by MuchosChuchos — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java (hunk 27 above)
Comment (Anna-Kudiakova) [reply api/src/test/java/com/eleks/academy/pharmagator/controllers/PriceControllerIT.java]: fixed

**Review by SofiiaOlenyn — APPROVED** [review]

**Review by MuchosChuchos — APPROVED** [review]

**Comment by Oleksandr-Eleks** [comment]
@Mykhailo-Cholii and @sergvasylchak , please approve if You have no more comments. Ready to merge! ;)

**Review by sergvasylchak — APPROVED** [review]

**Pull request merged**